segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/DefragmentOperation.java

```diff
@@ -0,0 +1,282 @@
+ * We might be able to give more knobs to tune its parameters (Eg. threshold on number of chunks).
+ * </li>
+ * <li>
+ * <li>
+ * Defrag operation will respect max rolling size and will not create chunks greater than that size.
+ * </li>
+ * </ul>
+ *
+ * What controls whether we invoke concat or simulate through appends?
+ * There are a few different capabilities that ChunkStorage needs to provide.
+ * <ul>
+ * <li>Does ChunkStorage support appending to existing chunks? For vanilla S3 compatible this would return false.
+ * This is indicated by supportsAppend.</li>
+ * <li>Does ChunkStorage support for concatenating chunks ? This is indicated by supportsConcat.
+ * If this is true then concat operation will be invoked otherwise chunks will be appended.</li>
+ * <li>There are some obvious constraints - For ChunkStorage support any concat functionality it must support either
+ * append or concat.</li>
+ * <li>Also when ChunkStorage supports both concat and append, ChunkedSegmentStorage will invoke appropriate method
+ * depending on size of target and source chunks. (Eg. ECS)</li>
+ * </ul>
+ *
+ * <li>
+ * What controls defrag?
+ * There are two additional parameters that control when concat
+ * <li>minSizeLimitForConcat: Size of chunk in bytes above which it is no longer considered a small object.
+ * For small source objects, append is used instead of using concat. (For really small txn it is rather efficient to use append than MPU).</li>
+ * <li>maxSizeLimitForConcat: Size of chunk in bytes above which it is no longer considered for concat. (Eg S3 might have max limit on chunk size).</li>
+ * In short there is a size beyond which using append is not advisable. Conversely there is a size below which concat is not efficient.(minSizeLimitForConcat )
+ * Then there is limit which concating does not make sense maxSizeLimitForConcat
+ * </li>
+ * <li>
+ * What is the defrag algorithm
+ * <pre>
+ * While(segment.hasConcatableChunks()){
+ *     Set<List<Chunk>> s = FindConsecutiveConcatableChunks();
+ *     For (List<chunk> list : s){
+ *        ConcatChunks (list);
+ *     }
+ * }
+ * </pre>
+ * </li>
+ * </ul>
+ */
+class DefragmentOperation implements Callable<CompletableFuture<Void>> {
+    private final MetadataTransaction txn;
+    private final SegmentMetadata segmentMetadata;
+    private final String startChunkName;
+    private final String lastChunkName;
+    private final ArrayList<String> chunksToDelete;
+    private final ChunkedSegmentStorage chunkedSegmentStorage;
+
+    private volatile ArrayList<ChunkInfo> chunksToConcat = new ArrayList<>();
+
+    private volatile ChunkMetadata target;
+    private volatile String targetChunkName;
+    private volatile boolean useAppend;
+    private volatile long targetSizeAfterConcat;
+    private volatile String nextChunkName;
+    private volatile ChunkMetadata next = null;
+
+    private volatile long writeAtOffset;
```

Comment:
Please try to either make these final or pass them along as arguments to your method calls.

Reply:
This should not matter as much, I used operation class so that I don't have to pass lots of parameters around. passing parameters becomes troublesome specially when multiple values need to be returned. These field are modified and then used in next step.